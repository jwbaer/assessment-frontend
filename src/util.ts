import { Component, List, PageApiResult, PageSpec, Variable, VariableDeclaration } from "./types";

export const convertPageApiResultToPageSpec = (apiResult: PageApiResult): PageSpec => {
    const componentIndex: Map<number, Component> = new Map(
        apiResult.components.map((obj: Component) => [obj.id, obj])
    );
    const listIndex: Map<number, List> = new Map(
        apiResult.lists.map((obj: List) => [obj.id, obj])
    );
    const variableIndex: Map<string, Variable> | undefined = apiResult.variables
        ? new Map(
            apiResult.variables.map((obj: VariableDeclaration) => [
                obj.name,
                { name: obj.name, type: obj.type, value: obj.initialValue },
            ])
        )
        : undefined;

    return({ componentIndex, listIndex, variableIndex });
}
