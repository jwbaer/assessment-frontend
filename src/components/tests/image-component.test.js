import { render, screen } from "@testing-library/react";

import ComponentPanel from "../component-panel";
import ImageComponent from "../image-component";
import PageContext from "../../contexts/page-context";
import { convertPageApiResultToPageSpec } from "../../util";

describe("<ImageComponent /> spec", () => {
    const apiResult = {
        lists: [
            {
                id: 0,
                components: [1],
            },
        ],
        components: [
            {
                id: 1,
                type: "image",
                options: {
                    src: "/locations/new-york.png",
                    alt: "Cartoon of New York skyline",
                },
            },
        ],
    };

    const pageSpec = convertPageApiResultToPageSpec(apiResult);

    it("renders the component", () => {
        const setPageSpec = jest.fn();
        const { asFragment } = render(
            <ImageComponent
                component={{
                    id: 1,
                    type: "image",
                    options: {
                        src: "/locations/new-york.png",
                        alt: "Cartoon of New York skyline",
                    },
                }}
            />
        );

        const image = screen.getByAltText("Cartoon of New York skyline");
        expect(image).toBeDefined();
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders when specifed through the component panel", () => {
        const setPageSpec = jest.fn();
        const { asFragment } = render(
            <PageContext.Provider value={{ pageSpec, setPageSpec }}>
                <ComponentPanel componentId={1} />
            </PageContext.Provider>
        );

        const image = screen.getByAltText("Cartoon of New York skyline");
        expect(image).toBeDefined();
        expect(asFragment()).toMatchSnapshot();
    });
});
