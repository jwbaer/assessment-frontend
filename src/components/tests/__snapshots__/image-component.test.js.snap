// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImageComponent /> spec renders the component 1`] = `
<DocumentFragment>
  <div
    class="component image"
    data-testid="image-component-1"
  >
    <img
      alt="Cartoon of New York skyline"
      data-testid="image-component-1-image"
      src="/locations/new-york.png"
    />
  </div>
</DocumentFragment>
`;

exports[`<ImageComponent /> spec renders when specifed through the component panel 1`] = `
<DocumentFragment>
  <div
    class="component image"
    data-testid="image-component-1"
  >
    <img
      alt="Cartoon of New York skyline"
      data-testid="image-component-1-image"
      src="/locations/new-york.png"
    />
  </div>
</DocumentFragment>
`;
